packages/core/src/stateUtils.ts

```diff
@@ -259,15 +259,51 @@ export const isStateId = (str: string) => str[0] === STATE_IDENTIFIER;
 
 export function getCandidates<TEvent extends EventObject>(
   stateNode: StateNode<any, TEvent>,
-  eventName: TEvent['type'] | NullEvent['type'] | '*'
+  eventName: TEvent['type'] | NullEvent['type'],
+  /**
+   * If `true`, will use SCXML event token matching semantics
+   */
+  matchTokens: boolean = false
 ): Array<TransitionDefinition<any, TEvent>> {
   const transient = eventName === NULL_EVENT;
   const candidates = stateNode.transitions.filter((transition) => {
-    const sameEventType = transition.eventType === eventName;
-    // null events should only match against eventless transitions
-    return transient
-      ? sameEventType
-      : sameEventType || transition.eventType === WILDCARD;
+    // First, check the trivial case: event names are exactly equal
+    if (transition.eventType === eventName) {
+      return true;
+    }
+
+    // transient transitions can't match non-transient events
+    if (transient) {
+        return false;
+    }
+    // Then, check if transition is a wildcard transition,
+    // which matches any non-transient events
+    if (transition.eventType === WILDCARD) {
+      return true;
+    }
+
+    if (!matchTokens && !transition.eventType.endsWith('.*')) {
+      return false;
+    }
+
+    const partialEventTokens = transition.eventType.split('.');
+    const eventTokens = eventName.split('.');
+
+    // tslint:disable-next-line: forin
+    for (const tokenIndex in partialEventTokens) {
+      const partialEventToken = partialEventTokens[tokenIndex];
+      const eventToken = eventTokens[tokenIndex];
+
+      if (partialEventToken === '*') {
```

Comment:
According to SCXML `.*` is only allowed at the end of the event descriptor and this would return `true` for a wildcard at any position in the delimited descriptor. Even if we don't validate the inputs and assume they are valid it still would be, IMHO, preferable not to trip on edge cases like this.

You can test this with:
```js
var matchTokens = true // just for simplicity of event descriptor etc, a similar effect can be achieved with false
var eventName = 'foo.xyz.baz'
var stateNode = { transitions: [{ eventType: 'foo.*.bar' }] }
```

Even if this would be a valid descriptor this `eventName` should not be matched by it.

Reply:
The `transition.eventType.endsWith('.*')` clause takes care of this.

Reply:
Added tests for this and disallowed the invalid event with a warning.